File: examples/validate_dots_data.py
import numpy

from mworks.data import MWKFile


def validate_dots_data(filename):
    with MWKFile(filename) as f:
        num_events = 0
        for e in f.get_events_iter(codes=['#announceStimulus']):
            value = e.value
            if (isinstance(value, dict) and
                (value['type'] == 'moving_dots') and
                ('dots' in value)):
                data = numpy.fromstring(value['dots'], numpy.float32)
                assert len(data) == (2 * value['num_dots'])
                
                x = data[::2] - value['field_center_x']
                y = data[1::2] - value['field_center_y']
                field_radius = value['field_radius']
                assert (x*x + y*y).max() <= (field_radius * field_radius)
                
                num_events += 1

        assert num_events > 0
        print 'Processed %d events' % num_events


if __name__ == '__main__':
    import sys
    validate_dots_data(sys.argv[1])

File: examples/validateDotsData.m
function validateDotsData(filename)

codec = getCodecs(filename);
codec = codec.codec;
all_codes = [codec(:).code];

for i = 1:length(codec)
    if strcmp(codec(i).tagname, '#announceStimulus')
        event_codes = [codec(i).code];
        break;
    end
end

events = getEvents(filename, event_codes);
nEvents = 0;

for i = 1:length(events)
    value = events(i).data;
    if isstruct(value) && strcmp(value.type, 'moving_dots') ...
            && isfield(value, 'dots')
        data = typecast(value.dots, 'single');
        assert(length(data) == (2 * value.num_dots), ...
               'Wrong number of dots');
        
        x = data(1:2:end) - value.field_center_x;
        y = data(2:2:end) - value.field_center_y;
        field_radius = value.field_radius;
        assert(max(x.*x + y.*y) <= (field_radius*field_radius), ...
               'Invalid dot coordinates');
        
        nEvents = nEvents + 1;
    end
end

assert(nEvents > 0, 'No events processed');
disp(sprintf('Processed %d events', nEvents));
